tools/src/main/java/org/apache/kafka/tools/GetOffsetShell.java

```diff
@@ -0,0 +1,404 @@
+        }
+
+        public boolean hasCommandConfigOpt() {
+            return options.has(commandConfigOpt);
+        }
+
+        public String commandConfigOpt() {
+            return options.valueOf(commandConfigOpt);
+        }
+
+        public String effectiveBrokerListOpt() {
+            return options.valueOf(effectiveBrokerListOpt);
+        }
+
+        public boolean hasExcludeInternalTopicsOpt() {
+            return options.has(excludeInternalTopicsOpt);
+        }
+    }
+
+    public Map<TopicPartition, Long> fetchOffsets(GetOffsetShellOptions options) throws IOException, ExecutionException, InterruptedException, TerseException {
+        String clientId = "GetOffsetShell";
+        String brokerList = options.effectiveBrokerListOpt();
+
+        if (options.hasTopicPartitionsOpt() && (options.hasTopicOpt() || options.hasPartitionsOpt())) {
+            throw new TerseException("--topic-partitions cannot be used with --topic or --partitions");
+        }
+
+        boolean excludeInternalTopics = options.hasExcludeInternalTopicsOpt();
+        OffsetSpec offsetSpec = parseOffsetSpec(options.timeOpt());
+
+        TopicPartitionFilter topicPartitionFilter;
+
+        if (options.hasTopicPartitionsOpt()) {
+            topicPartitionFilter = createTopicPartitionFilterWithPatternList(options.topicPartitionsOpt());
+        } else {
+            topicPartitionFilter = createTopicPartitionFilterWithTopicAndPartitionPattern(
+                    options.hasTopicOpt() ? options.topicOpt() : null,
+                    options.partitionsOpt()
+            );
+        }
+
+        Properties config;
+
+        if (options.hasCommandConfigOpt()) {
+            config = Utils.loadProps(options.commandConfigOpt());
+        } else {
+            config = new Properties();
+        }
+
+        config.setProperty(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList);
+        config.setProperty(AdminClientConfig.CLIENT_ID_CONFIG, clientId);
+
+        try (Admin adminClient = Admin.create(config)) {
+            List<TopicPartition> partitionInfos = listPartitionInfos(adminClient, topicPartitionFilter, excludeInternalTopics);
+
+            if (partitionInfos.isEmpty()) {
+                throw new TerseException("Could not match any topic-partitions with the specified filters");
+            }
+
+            Map<TopicPartition, OffsetSpec> timestampsToSearch = partitionInfos.stream().collect(Collectors.toMap(tp -> tp, tp -> offsetSpec));
+
+            ListOffsetsResult listOffsetsResult = adminClient.listOffsets(timestampsToSearch);
+
+            TreeMap<TopicPartition, Long> partitionOffsets = new TreeMap<>(Comparator.comparing(TopicPartition::toString));
+
+            for (TopicPartition partition : partitionInfos) {
+                ListOffsetsResultInfo partitionInfo;
+
+                try {
+                    partitionInfo = listOffsetsResult.partitionResult(partition).get();
+                } catch (KafkaException e) {
```

Comment:
I don't think this throws `KafkaException`. This throws `ExecutionException` and you need to check the cause exception instead.